File: data_struct/list.cpp
#include <iostream>
#include "list.h"

using namespace std;


//Constructor
List::List(){
    head=NULL;
    curr=NULL;
    temp=NULL;
}


//List length
int List::listLen(){
    int len=0;
    curr=head;
    while(curr->next != NULL){
        len++;
        curr=curr->next;
    }
    len++;
    return len;
}


//Print Data
void List::printList(){
    if(head!=NULL){
        curr=head;
        while(curr->next!=NULL){
            cout<<"|"<<curr->data;
            curr = curr->next;
        }
        cout<<"|"<<curr->data<<"|"<<endl;
    }
    else
        cout<<"Empty List"<<endl;
}


//Insert Data
void List::insNode(int newData, int pos){
    nodePtr n = new Node();
    n->data = newData;
    n->next = NULL;
    
    //Insert at the start
    if(pos == 0){
        if(head != NULL){
            curr=head;
            head=n;
            n->next = curr;
        }
        else
            head=n;
    }
    //Insert at the end
    else if(pos == -1){
        if(head != NULL){
            curr=head;
            while(curr->next != NULL)
                curr = curr->next;
            curr->next = n;
        }
        else
            head=n;
    }
    //Insert at a position
    else{
        int len = listLen();
        if(pos <= len){
            int posCount=0;
            curr=head;
            while(posCount != pos){
                temp = curr;
                curr = curr->next;
                posCount++;
            }
            temp->next = n;
            n->next=curr;
        }
        else
            cout<<"Given node out of range"<<endl;
    }   
}

void List::delNode(int pos){
    //Remove the first node
    if(pos == 0){
        curr=head;
        head = curr->next;
        delete curr;
    }
    else if(pos == -1){
        curr=head;
        while(curr->next != NULL){
            temp = curr;
            curr = curr->next;            
        }
        temp->next = NULL;
        delete curr;
    }
    else{
        int len = listLen();
        if(pos <= len){
            int posCount = 0;
            curr = head;
            while(posCount != pos){
                temp =  curr;
                curr = curr->next;
                posCount++;
            }
            temp->next = curr->next;
            delete curr;
        }
        else
            cout<<"Given node out of range"<<endl;
    }
}

//Returns the pointer of a node
nodePtr List::getNode(int pos){
    if(pos == 0)
        return head;
    else if(pos == -1){
        curr = head;
        while(curr->next != NULL)
            curr = curr->next;
        return curr;
    }
    else{
        if(pos <= listLen()){
            int posCount = 0;
            while(posCount < pos){
                curr = curr->next;
                posCount++;
            }
            return curr;
        }
        else{
            cout<<"Given node out of range"<<endl;
            return NULL;
        }
    }
    
}

File: pyads/list.py
class Node:
    def __init__(self, data, nextNode=None):
        self.value = data
        self.nextNode = nextNode

class SingleLL:
    def __init__(self):
        self.head = None
        
    def prepend(self, data):
        n = Node(data)
        if self.head == None:
            self.head = n
        else:
            temp = self.head
            self.head = n
            n.nextNode = temp
    
    def append(self, data):
        n = Node(data)
        if self.head == None:
            self.head = n
        else:
            curr = self.head
            while(curr.nextNode != None):
                curr = curr.nextNode
            curr.nextNode = n
    
    def printList(self):
        curr = self.head
        while(curr.nextNode != None):
            print(curr.value)
            curr = curr.nextNode
        print(curr.value)
        
    def delFirstNode(self):
        if self.head == None:
            print("Empty List")
        else:
            curr = self.head
            self.head = curr.nextNode
            del curr
    
    def delLastNode(self):
        if self.head == None:
            print("Empty List")
        else:
            curr = self.head
            while(curr.nextNode != None):
                temp = curr
                curr = curr.nextNode
            temp.nextNode = None
            del curr
